---
title: Generating Correlated Data for Omics Simulation
author:
    - name: "Jianing Yang"
      orcid: 0000-0002-2048-9398
      affiliations:
        - name: "Chronobiology and Sleep Institute, University of Pennsylvania"
    - name: "Gregory R. Grant"
      orcid: 0000-0002-0139-7658
      affiliations:
        - name: "Department of Genetics, University of Pennsylvania"
    - name: "Thomas G. Brooks"
      orcid: 0000-0002-6980-0079
      affiliations:
        - name: "Institute for Translational Medicine and Therapeutics, University of Pennsylvania"
editor: 
  markdown: 
    wrap: sentence
execute:
  echo: false
  cache: true
  warning: false
format:
  html:
    toc: true
    toc-expand: true
bibliography: ../references.bib
---

## Supplemental Material

### Comparisons to real data

```{r}
#| label: fig-compare-to-real
#| fig-cap: "Comparison to real data run on a fly whole body data set from GSE81142. (a-b) Comparison of gene (a) mean expression and (b) variance, log-scaled in real and PCA simulated data. The line of equality is marked in black. Points are colored according to the density of points in their region. Wishart and corpcor methods give similar results (not shown). (c) Quantile-quantile plot comparing correlation values of gene pairs from real data and simulated data (both with and without dependence). Genes with at least 30 reads were used. Values on the diagonal line indicate a match between the simulated and real data sets. (d) Projections onto the top two principal components of the real data set for both real and simulated data. All 8 simulations (96 samples for each simulation) shown. (e) Principal component analysis was performed on all data sets and the variance captured by the top components is shown. Unlike (d), these components were fit from each data set considered separately instead of reusing the weights from the real data."
#| fig-width: 10
#| fig-height: 10
readRDS("../../processed/compare_to_real_plot.GSE81142.RDS")
```

```{r}
#| label: fig-compare-to-real2
#| fig-cap: "Comparison to real data run on a mouse cortex timeseries body data set from GSE151565. Only timepoint ZT0 was used for this comparison. (a-b) Comparison of gene (a) mean expression and (b) variance, log-scaled in real and PCA simulated data. The line of equality is marked in black. Points are colored according to the density of points in their region. Wishart and corpcor methods give similar results (not shown). (c) Quantile-quantile plot comparing correlation values of gene pairs from real data and simulated data (both with and without dependence). Genes with at least 30 reads were used. Values on the diagonal line indicate a match between the simulated and real data sets. (d) Projections onto the top two principal components of the real data set for both real and simulated data. All 8 simulations (96 samples for each simulation) shown. (e) Principal component analysis was performed on all data sets and the variance captured by the top components is shown. Unlike (d), these components were fit from each data set considered separately instead of reusing the weights from the real data."
#| fig-width: 10
#| fig-height: 10
readRDS("../../processed/compare_to_real_plot.GSE151565.RDS")
```

## Supplemental Methods

### Running SPsimSeq

The R package SPsimSeq [@Assefa2020] provides a dedicated RNA-seq and single-cell RNA-seq simulator using a Gaussian copula approach to simulate gene dependence.
We ran SPsimSeq using the following options:

``` R
    n.sim = 1,
    s.data = input,
    n.genes = input |> nrow(),
    batch.config = 1,
    group.config = 1,
    pDE = 0,
    tot.samples = 100,
    model.zero.prob = FALSE,
    genewiseCor = TRUE,
    log.CPM.transform = TRUE,
    result.format = "list",
    return.details = TRUE,
    verbose = TRUE,
```

This produces bulk RNA-seq data with no batching or group effects or differentially expressed (DE) genes.
Raw count values provided as the input matrix.
For dataset GSE151565, only the ZT0 timepoints were used.
This is the same set of samples that were used in our simulation to determine the correlation structure (copula).


SPsimSeq was not run in the two example applications, DESeq2 and CYCLOPS, due to substantial differences in how it generates DE across samples compared to our method.
DE in SPsimSeq is done by mimicking a random set of detected DE genes in the reference data set.
SPsimSeq does not report either the group mean values or the log fold changes used for these genes, so we are not able to modify our own simulation to have the same set of DE genes at the same effect sizes.

### Running vine copula

We used the rvinecopulib package to fit and generate random data from a vine copula distribution.
This package is a R wrapper of the high-performance vinecopulib C++ package.
This package expects data scaled to the range of \[0, 1\].
Therefore, we chose not to use the real RNA-seq count data and instead generated correlated random number as input which follow a multivariate normal distribution with covariance matrix constructed with a randomly generated singular vector matrix and singular values 1, 2, ..., 100.
These values were then scaled to the interval \[0,1\] by the `pnorm` function.
A vine copula was fit using just the "guassian" family.
While rvinecopula supports multi-threaded operation, we ran using a single thread as with the other methods.

### Running mvrnorm

To assess a simple multivariate normal distribution simulation option, we ran the `mvrnorm` function from the MASS R package.
We set means to be the mean expression of each gene and used the sample covariance matrix as the `Sigma` parameter.
