---
title: Generating Correlated Data for Omics Simulation
editor: 
  markdown: 
    wrap: sentence
execute:
  echo: false
  cache: true
format:
  html:
    toc: true
    toc-expand: true
bibliography: references.bib
---

## Abstract

Simulation of realistic omics data is a key input for benchmarking studies that help users obtain optimal computational pipelines.
Omics data involves large numbers of measured features on each samples and these measures are generally correlated with each other.
However, simulation too often ignores these correlations, perhaps due to the inconvenience and computational hurdles of doing so.
We describe two approaches to solving this problem that are simple to implement and computationally very fast, both of which work by decomposing the covariance matrix into a diagonal part and a low-rank part.
We demonstrate the importance of including correlation in two benchmarking applications.
First, we show that variance of results from the popular DESeq2 method increases when dependence is included.
Second, we demonstrate that a method, CYCLOPS, for inferring circadian time of collection from transcriptomics actually improves its performance when dependence is included.
We provide an R package, dependentsimr, that has efficient implementations of these methods and can generate dependent data with arbitrary distributions, including discrete (binary, ordered categorical, Poisson, negative binomial), continuous (normal), or with an empirical distribution.

## Introduction

Omics data is in the “p \>\> n” regime where there are fewer samples than measurements per sample.
This creates dual challenges in generating realistic simulated data for the purposes of benchmarking.
First, there isn’t enough data to be able to compute a dependence structure (e.g., a full-rank correlation matrix).
Second, generating omics-scale data with a specified correlation matrix is slow due to the typical $O(p^3)$ nature of these algorithms.
These often mean that simulators assume independence of the measurements, which does not reflect reality.

Recently, we wrote guidelines for performing omics benchmarking [@Brooks2024] and encouraged authors to ensure that any simulated data includes realistic correlation and dependence of measurements within a sample.
However, this highlighted that one reason this is often neglected is that there is relatively little guidance for how to achieve it and that existing solutions are cumbersome.

Here, we give a simple solution this problem by using a correlation matrix that decomposes into a diagonal part and a low-rank part to both approximate realistic dependencies in a real data set and generate simulated data mimicking the real data.
Using a NORTA (Normal to Anything) approach, the marginal (univariate) distributions can have realistic forms.
We implement this strategy as an R package which supports normal, Poisson, DESeq2-based (negative binomial with sample-specific size factors), and empirical (for ordinal data) marginal distributions.
This makes it particularly suited to RNA-seq data but also widely applicable.

We show two applications which demonstrate the importance of including dependence of measurements in simulated data when benchmarking computational pipelines.
First, we simulate RNA-seq data with differential expression between two conditions.
Using DESeq2 to determine the differentially expressed genes, we found that dependence had little impact on the accuracy of reported p-values but increased the variance of those estimates, meaning that experiments simulated with dependence were more likely to be conservative and more likely to be anti-conservative compared to simulations of independent data.
Second, we simulated a time series of RNA-seq data points and used the CYCLOPS method to infer collection time from the RNA-seq data, without time labels.
Surprisingly, we found that including dependence in the data set resulted in improved performance with CYCLOPS.

## Results

### Comparison to Real Data

```{r}
#| include: false
source("compare_to_real.R")
```

We used mouse liver RNA-seq data set from accession GSE77221[@Weger2019] where we expected a significant amount of gene-gene correlation due to the samples having been collected throughout the course of a day.
We used this data to prime our simulation and compared real and simulated data sets, both with 12 samples, to each other.
We repeated the simulations a total of 8 times to estimate variance.
The simulations were then repeated without any dependence for comparison.

Simulated data captures the genes' mean and variance accurately (@fig-compare-to-real a-b).
Moreover, we computed the gene-gene correlation on a random subset of `r N_rows_gene_corr` high-expressed genes.
Since only 12 samples are in each real or simulated data set, correlation estimates are very noisy.
However, the distribution of correlations across all pairs of these genes is much closer to that of real data when simulated with dependence than without (@fig-compare-to-real c).

Next, we compared the simulations with and without dependence to the real data set when projected onto the top two principal components of the real data set (@fig-compare-to-real d).
The simulations with dependence are distributed around the entire space like the real data, but the independent simulations have unrealistically low variance in these components, clustering tightly around the origin.
Lastly, we compared the PCA of each simulated data set considered separately from the real data.
Simulations with independent data showed unrealistically low levels of variance in top principal components, but dependent simulations actually show too high elevations of variance in top components.
This is a consequence of our PCA strategy: since variance in the direction of the real data set's top principal component is exactly matched, the simulated data set's top principal component will be even larger than that.

```{r}
#| label: fig-compare-to-real
#| fig-cap: !expr glue::glue("Comparison to real data run on a mouse liver example from GSE77221. (a-b) Comparison of gene (a) mean expression and (b) variance, log-scaled. The line of equality is marked in black. Points are colored according to the density of points in their region. (c) Quantile-quantile plot comparing correlation values of gene pairs from real data and simulated data (both with and without dependence). A random sample of {N_rows_gene_corr} genes with at least {HIGH_EXPR_CUTOFF} reads was used. Values on the diagonal line indicate a match between the simulated and real data sets. (d) Projections onto the top two principal components of the real data set for both real and simulated data. All 8 simulations (96 samples total for each of dependent and independent simulations) shown. (e) Principal component analysis was performed on all data sets and the variance captured by the top components is shown. Unlike (d), these components were fit from each data set considered separately instead of reusing the weights from the real data.")
#| fig-width: 10
#| fig-height: 10
compare_to_real_plot
```

### DESeq2 Application

We benchmark DESeq2, a popular differential expression analysis tool, using data sets simulated with dependence and ones simulated without dependence to compare its performances on both. We use a fly whole body RNA-Seq data set GSE81142 and select samples of male flies without treatment and after at least 2 hours of feeding to simulate 20 control samples with rank $k=2$ dependence structure, and 20 control samples assuming independence of the genes. We then randomly select 5% of the genes to be up or down regulated, with $\log_2$ fold change uniformly distributed between $0.2$ and $2.0$, and simulate 20 "experimental" samples each with and and without dependence.

Finally, we run DESeq2 on the two simulated experiments and compare the output false discovery rate (FDR) with the true percentages of genes that are differential expressed (@fig-DESeq2 a-b). We observe that DESeq2 is slightly anti-conservative on both data sets, with similar mean true FDRs for each estimated FDR cutoff. However, there is a greater variance in the performance of DESeq2 on the data sets simulated with dependence, indicating that it potentially performs less reliably on data sets with gene-gene dependence, as in real data sets. We speculate that this is due to DESeq2's assumption that the gene expressions are independent.

To demonstrate the application of our simulation method for another organism, we also simulate data sets using a mouse cortex data set GSE151923 and select samples from male mice to infer the dependence structure of the quantified genes. We then simulate differential expression experiments with and without dependence as above, with 20 samples for each condition. We observe a similar result (@fig-DESeq2 c-d) as for the fly whole body data sets.

```{r}
#| include: false
source("DE_plots.R")
```

```{r}
#| label: fig-DESeq2
#| fig-cap: "Performance of DESeq2 on simulated datasets. (a-b) Comparison of true FDR and DESeq2 reported FDR for data sets simulated from the fly whole body data set (GSE81142), (a) without dependence and (b) with dependence. (c-d) Comparison of true FDR and DESeq2 reported FDR for data sets simulated from the mouse cortex data set (GSE151923, Wang et al. 2022), (c) without dependence and (d) with dependence."
#| fig-width: 7
#| fig-height: 6
DESeq2_fdr_plot
```

### CYCLOPS Application

We further use our simulation method to benchmark CYCLOPS (Anafi et al. 2017), which infer relative times for a set of unlabeled samples using an autoencoder to identify circular structures. We use a mouse liver time series data set (GSE151565, Aaronson et al. 2020), which contains a total of 77 samples every 3 hours, for 36 hours. We compute the dependence structure of the genes as well as the variances of marginal distributions using the 12 time point 0 samples, and compute the means of gene expressions at each time point. We then use these to simulate 20 time series data sets with rank $2$ dependence structure, each with 8 time points and 6 samples per time point, and 20 such data sets without dependence.

We run CYCLOPS on each data set with a list of cyclic mouse liver genes (JTK p-value $<0.05$), which yields an estimated relative time for each sample. For the evaluation metric for CYCLOPS' performance, we use the circular correlation, defined as follows:
$$\rho = \frac{\sum_{1\leq i<j\leq n}sin(X_i-X_j)sin(Y_i-Y_j)}{(\sum_{1\leq i<j\leq n}sin(X_i-X_j)^2)^{1/2}(\sum_{1\leq i<j\leq n}sin(Y_i-Y_j)^2)^{1/2}},$$
where $n$ is the number of samples, $X_i$ and $Y_i$ are the true time and CYCLOPS-estimated time respectively for the $i$-th sample. $\rho$ has value between $-1$ and $1$, and a $|\rho|$ close to $1$ indicates accurate predictions by CYCLOPS.

We find that CYCLOPS performs better overall on simulated time series data sets with dependence than on those without dependence (@fig-CYCLOPS), indicating that the presence of gene-gene dependence is essential for CYCLOPS' performance.

```{r}
#| include: false
source("circular_correlation.R")
```

```{r}
#| label: fig-CYCLOPS
#| fig-cap: "Performance of CYCLOPS on simulated time series data sets based on mouse liver data set (GSE151565, Aaronson et al. 2020). (a) Violin plot comparing absolute circular correlations between true phases and CYCLOPS estimated phases on the simulated data sets with and without dependence. (b-c) Examples of CYCLOPS estimated phases on a simulated data set (b) without dependence and (c) with dependence."
cyclops_plot
```

## Methods

TODO: how it works

## Discussion

We present `dependentsimr`, an R package that generates omics-scale data with realistic correlation.
TODO: limitations TODO: alternatives
